cmd/argocd/commands/app.go

```diff
@@ -1291,25 +1336,30 @@ func waitOnApplicationStatus(acdClient apiclient.Client, appName string, timeout
 			refresh = true
 		}
 
-		healthStatus := true
-		if watchSuspened && watchHealth {
-			healthStatus = app.Status.Health.Status == argoappv1.HealthStatusHealthy ||
-				app.Status.Health.Status == argoappv1.HealthStatusSuspended
-		} else if watchSuspened {
-			healthStatus = app.Status.Health.Status == argoappv1.HealthStatusSuspended
-		} else if watchHealth {
-			healthStatus = app.Status.Health.Status == argoappv1.HealthStatusHealthy
-		}
+		// If selected resources are included, wait only on those resources, otherwise wait on the application as a whole.
+		if len(selectedResources) > 0 {
+			selectedResourcesAreReady := true
+			for _, state := range getResourceStates(app, selectedResources) {
+				resourceIsReady := checkResourceStatus(watchSync, watchHealth, false, watchSuspended, state.Health, state.Status, nil)
+				if !resourceIsReady {
+					selectedResourcesAreReady = false
+				}
+			}
 
-		// consider skipped checks successful
-		synced := !watchSync || app.Status.Sync.Status == argoappv1.SyncStatusCodeSynced
-		operational := !watchOperation || appEvent.Application.Operation == nil
-		if len(app.Status.GetErrorConditions()) == 0 && synced && healthStatus && operational {
-			printFinalStatus(app)
-			return app, nil
+			if selectedResourcesAreReady {
+				printFinalStatus(app)
+				return app, nil
+			}
+
+		} else {
+			applicationIsReady := checkResourceStatus(watchSync, watchHealth, watchOperation, watchSuspended, app.Status.Health.Status, app.Status.Sync.Status, appEvent.Application.Operation)
+			if len(app.Status.GetErrorConditions()) == 0 && applicationIsReady {
```

Comment:
I think we should use this GetErrorConditions check with individual resources too since the individual resources may be in an odd state if there are any errors. @jessesuen what do you think?

Reply:
On a closer look, I agree too. I think we should not tell the user that their resources are healthy if the application itself has an error condition. I will add it to the code.